Add PyObjectSlot member to StorageImpl

kurtamohler opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93342

Part of #91395

Also modifies how `StorageImpl`s are stored in JIT static runtime's `MemoryPlanner`, which used to `std::move` `StorageImpl`s into a vector. But `StorageImpl` can no longer be moved. Instead, `MemoryPlanner` now contains a malloced buffer to which we add new `StorageImpl`s using placement new

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

kurtamohler committed:
Add PyObjectSlot member to StorageImpl

[ghstack-poisoned]

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

ezyang commented on torch/csrc/StorageSharing.cpp:
```diff
@@ -90,34 +90,43 @@ static PyObject* THPStorage_shareFilename(PyObject* _self, PyObject* noargs) {
+    // Replace the old data_ptr and allocator with the new ones
+    c10::StorageImpl* new_storage_impl = new_storage.unsafeGetStorageImpl();
+    self_storage_impl->set_data_ptr(std::move(new_storage_impl->data_ptr()));
+    self_storage_impl->set_allocator(new_storage_impl->allocator());
```
nice

Skylion007 commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -169,6 +169,7 @@ class MemoryPlanner {
     // UB. We're doing fancy memory allocation stuff, so we cast to an
     // integer type and carry on.
     const auto impl_p = reinterpret_cast<uintptr_t>(impl);
+    // TODO: I think this might not work after changing to intrusive_ptr?
```
Just dereference the smart ptr?

kurtamohler commented:
I'm trying to build the benchmarks in `benchmarks/static_runtime` to reproduce the CI errors. There don't seem to be any build instructions, and this doesn't work for me:

```shell
$ cd benchmarks/static_runtime
$ cmake .
$ make all
```

It generates some cmake files, but doesn't build any of the binaries

ezyang commented on c10/core/StorageImpl.h:
```diff
@@ -222,5 +223,6 @@ struct C10_API StorageImpl : public c10::intrusive_ptr_target {
   // local to process cuda memory allocation
   bool received_cuda_;
   Allocator* allocator_;
+  impl::PyObjectSlot pyobj_slot_;
```
@swolchok gonna be so mad 🥲 (this adds two words per storage which is not nothing, but I don't really know how else we can do this PR set)

swolchok replied:
are there platforms (like, say, `C10_MOBILE`) where we can statically identify that we don't need this?

kurtamohler replied:
Any platform that doesn't use Python should not need the PyObjectSlot. I'm not sure if there are any other than `C10_MOBILE`. I can add an ifdef if you'd like

ezyang commented on torch/csrc/jit/runtime/static/memory_planner.cpp:
```diff
@@ -375,14 +375,15 @@ void StandardMemoryPlanner::allocateManagedTensors() {
       continue;
     }
-    at::StorageImpl* storageImpl = &ms.second;
+    c10::intrusive_ptr<at::StorageImpl> storageImpl = ms.second;
```
This incurs a reference count bump

ezyang commented:
It doesn't look like you have to build it separately. It should be built in build/bin/static_runtime_test if you BUILD_TEST=1 when building PyTorch.

But let's talk about static runtime for a moment. They seem to really want the StorageImpl to be laid out contiguously without requiring an extra indirection (read the comment above). This is going to be difficult to arrange for because I told you to make StorageImpl non-movable. If they never resize the vector, we can reimplement this using a malloc'd buffer which we placement-new the StorageImpls into. Can you check if they do resize? If they don't, this may be cause to work out how to keep StorageImpl movable.

kurtamohler commented:
@ezyang, the vector in the static runtime memory planner is never resized and it looks like a malloced buffer would work. So should we make StorageImpl movable?

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

ezyang commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -199,7 +192,7 @@ class MemoryPlanner {
   // Storage for managed tensors out from under us during execution,
   // so we have to check the StorageImpls each time we deallocate.
-  std::vector<std::pair<size_t, at::StorageImpl>>
+  std::vector<std::pair<size_t, c10::intrusive_ptr<at::StorageImpl>>>
```
This doesn't have the same perf properties as the old code, alas

kurtamohler commented:
I've confirmed that the `managed_tensor_storage_impls_` vector's capacity never changes after the first storage is put into it. I first tested this by keeping an expected capacity variable and asserting that the capacity of the vector is always equal to it after the `emplace_back` call. I ran the `static_runtime_test` and `static_runtime_bench`, and they didn't throw an error, so the vector's capacity must be staying unchanged.

That was probably enough, but I also tried replacing the vector with two heap-allocated arrays just to make sure, and that worked fine too.

So should we go forward with enabling `StorageImpl` to be moved?

ezyang commented:
Nope, since the capacity never changes, you can do the malloc-inplace initialize trick.

kurtamohler commented:
Ok I misunderstood. So are you suggesting to basically copy the `StorageImpl`s into a buffer in the `MemoryPlanner`?

I don't think copying the `PyObjectSlot` would work at the moment because whichever `StorageImpl` gets destructed first will deallocate the `PyObject`. Maybe it's possible to make it work. But on the other hand, maybe the copies of the `StorageImpl`s can just have an uninitialized `PyObjectSlot`. I'm not sure if the static runtime needs the PyObject or not

ezyang commented:
I wouldn't use the word "copy" since that has a specific technical meaning in C++, but yes, we can construct a "vector" of StorageImpls that are contiguous by mallocing a buffer and then placement-new'ing each slot.

In the case of static runtime, I think the PyObject slots are always null (because this is a C++ only concept), so you don't have to worry about that interaction.

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -199,8 +192,25 @@ class MemoryPlanner {
+
+  // This must be called by the destructor of any subclass, so that the buffer
+  // of storages is deleted properly
+  void deleteStorages();
```
Maybe it would be better to put this storage buffer stuff into a separate class (maybe called `StorageImplsBuffer`) so that the logic for allocating and deleting is separate, and we wouldn't have to explicitly call this `deleteStorages` function

ezyang replied:
Yeah, a mini wrapper class is safer and highly recommended.

ezyang commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -199,8 +192,25 @@ class MemoryPlanner {
-      managed_tensor_storage_impls_{};
+
+  // We will use placement-new to add new storages to this buffer of bytes.
+  alignas(at::StorageImpl) unsigned char* storages_buffer_;
```
This does not do what you think it does. This will ensure that the char* POINTER is aligned to StorageImpl size in the class. But you need the malloc'ed memory to be aligned instead.

ezyang replied:
Actually, why does the malloc'ed memory need to be aligned, I don't think that's necessary either.

kurtamohler replied:
I see, it is not necessary. I didn't know what `alignas` does before, I was just copying from this and thought it was important here too: https://en.cppreference.com/w/cpp/language/new#Placement_new

But I can instead just use the one `StorageImpl* storages_` pointer and initialize it like `storages_ = reinterpret_cast<StorageImpl*>(new unsigned char[storages_capacity_ * sizeof(StorageImpl)])`, and that works just fine. (The `reinterpret_cast` is necessary because `StorageImpl::StorageImpl()` without args is not allowed.)

ezyang commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -199,8 +192,25 @@ class MemoryPlanner {
+  alignas(at::StorageImpl) unsigned char* storages_buffer_;
+
+  // This pointer just points to the storage buffer with the proper type
+  at::StorageImpl* storages_;
```
Having both pointers should not be necessary, only one should be necessary.

ezyang commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -199,8 +192,25 @@ class MemoryPlanner {
+  size_t storages_size_;
+
+  // Total capacity of the storage buffer
+  size_t storages_capacity_;
```
I'm not sure why you need all of these sizes; a simple count of how many storages are in the buffer feels sufficient

ezyang replied:
Oh I see the old code stored a pair. Store the size_t in a normal vector lol

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

swolchok commented on torch/csrc/jit/runtime/static/memory_planner.cpp:
```diff
@@ -364,19 +420,20 @@ void StandardMemoryPlanner::allocateManagedTensors() {
   auto group_idx = 0;
-  for (auto& ms : managed_tensor_storage_impls_) {
-    auto tensor_size = ms.first;
+  for (size_t storages_idx = 0; storages_idx < storages_.size();
```
irange

swolchok commented on torch/csrc/jit/runtime/static/memory_planner.cpp:
```diff
@@ -135,6 +135,62 @@ std::vector<StorageGroup> assignStorageToManagedTensors(
+  }
+}
+
+at::StorageImpl* ManagedStorages::operator[](size_t idx) {
+  TORCH_INTERNAL_ASSERT(storages_ != nullptr);
+  return &storages_[idx];
+}
+
+const at::StorageImpl* ManagedStorages::operator[](size_t idx) const {
+  TORCH_INTERNAL_ASSERT(storages_ != nullptr);
+  return &storages_[idx];
+}
+
```
please put trivial accessors like this (and size and capacity below) in the header so that they can easily be inlined without relying on LTO

swolchok commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -162,18 +194,16 @@ class MemoryPlanner {
 
   bool isManagedStorageImpl(const at::StorageImpl* impl) const {
-    if (managed_tensor_storage_impls_.empty()) {
+    if (storages_.size() == 0) {
```
I see a lot of 0 comparisons against size(); perhaps we should add `empty()`

swolchok commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -39,6 +39,38 @@ class StorageGroup {
+
+  void append(at::StorageImpl* storageImpl);
+
+  at::StorageImpl* operator[](size_t idx);
+
+  const at::StorageImpl* operator[](size_t idx) const;
```
wait, why do these return pointers instead of references?

kurtamohler replied:
No good reason, I've changed them to references

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

Also modifies how `StorageImpl`s are stored in JIT static runtime's `MemoryPlanner`, which used to `std::move` `StorageImpl`s into a vector. But `StorageImpl` can no longer be moved. Instead, `MemoryPlanner` now contains a malloced buffer to which we add new `StorageImpl`s using placement new

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler commented:
Thanks for the review, @swolchok! Let me know if there's anything else

ezyang commented on torch/csrc/jit/runtime/static/memory_planner.h:
```diff
@@ -39,6 +39,54 @@ class StorageGroup {
+
+  ~ManagedStorages();
+
+  void allocate(size_t capacity);
```
It's worth commenting that this invalidates all the old storages. In fact, if this is never used to reallocate, I might just say to make this an error to reallocate.

ezyang commented on torch/csrc/jit/runtime/static/memory_planner.cpp:
```diff
@@ -135,6 +135,44 @@ std::vector<StorageGroup> assignStorageToManagedTensors(
+  }
+  capacity_ = capacity;
+  storages_ = reinterpret_cast<at::StorageImpl*>(
+      new unsigned char[capacity_ * sizeof(at::StorageImpl)]);
```
You gonna set `size_`?

kurtamohler replied:
Setting `size_ = 0` here shouldn't be necessary. If `storages_ == nullptr`, then `size_` is already 0. Otherwise, `deallocate()` will set it to 0

kurtamohler replied:
But it would probably be a good idea to add an internal assert that `size_ == 0` just to make sure

ezyang reviewed (approved):
lgtm

eellison commented:
Hey, this looks great ! Would you mind landing it?

kurtamohler commented:
@pytorchbot merge

kurtamohler commented:
@pytorchbot rebase

kurtamohler commented:
I'm on PTO from now to March 7. I'll fix the conflict and land when I'm back

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

Also modifies how `StorageImpl`s are stored in JIT static runtime's `MemoryPlanner`, which used to `std::move` `StorageImpl`s into a vector. But `StorageImpl` can no longer be moved. Instead, `MemoryPlanner` now contains a malloced buffer to which we add new `StorageImpl`s using placement new

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

Also modifies how `StorageImpl`s are stored in JIT static runtime's `MemoryPlanner`, which used to `std::move` `StorageImpl`s into a vector. But `StorageImpl` can no longer be moved. Instead, `MemoryPlanner` now contains a malloced buffer to which we add new `StorageImpl`s using placement new

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

Also modifies how `StorageImpl`s are stored in JIT static runtime's `MemoryPlanner`, which used to `std::move` `StorageImpl`s into a vector. But `StorageImpl` can no longer be moved. Instead, `MemoryPlanner` now contains a malloced buffer to which we add new `StorageImpl`s using placement new

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler committed:
Update on "Add PyObjectSlot member to StorageImpl"


Part of #91395

Also modifies how `StorageImpl`s are stored in JIT static runtime's `MemoryPlanner`, which used to `std::move` `StorageImpl`s into a vector. But `StorageImpl` can no longer be moved. Instead, `MemoryPlanner` now contains a malloced buffer to which we add new `StorageImpl`s using placement new

cc ezyang bhosmer smessmer ljk53 bdhirsh

[ghstack-poisoned]

kurtamohler commented:
@pytorchbot merge -f "unrelated failure"
